File: app/src/main/java/me/devsaki/hentoid/parsers/images/BaseImageListParser.java
package me.devsaki.hentoid.parsers.images;

import android.webkit.URLUtil;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.util.Pair;

import com.annimon.stream.Optional;

import org.apache.commons.lang3.NotImplementedException;
import org.apache.commons.lang3.tuple.ImmutablePair;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;

import me.devsaki.hentoid.database.domains.Chapter;
import me.devsaki.hentoid.database.domains.Content;
import me.devsaki.hentoid.database.domains.ImageFile;
import me.devsaki.hentoid.enums.StatusContent;
import me.devsaki.hentoid.events.DownloadCommandEvent;
import me.devsaki.hentoid.parsers.ParseHelper;
import me.devsaki.hentoid.util.exception.EmptyResultException;
import me.devsaki.hentoid.util.exception.LimitReachedException;
import me.devsaki.hentoid.util.network.HttpHelper;
import timber.log.Timber;

public abstract class BaseImageListParser implements ImageListParser {

    private final ParseProgress progress = new ParseProgress();
    protected AtomicBoolean processHalted = new AtomicBoolean(false);
    protected String processedUrl = "";

    protected abstract boolean isChapterUrl(@NonNull String url);

    protected abstract List<String> parseImages(@NonNull Content content) throws Exception;

    protected abstract List<String> parseImages(@NonNull String chapterUrl, String downloadParams, List<Pair<String, String>> headers) throws Exception;

    public List<ImageFile> parseImageList(@NonNull Content onlineContent, @NonNull Content storedContent) throws Exception {
        return parseImageListImpl(onlineContent, storedContent);
    }

    public List<ImageFile> parseImageList(@NonNull Content content) throws Exception {
        return parseImageListImpl(content, null);
    }

    public List<ImageFile> parseImageList(@NonNull Content content, @NonNull String url) throws Exception {
        if (isChapterUrl(url)) return parseChapterImageListImpl(url, content);
        else return parseImageListImpl(content, null);
    }

    protected List<ImageFile> parseImageListImpl(@NonNull Content onlineContent, @Nullable Content storedContent) throws Exception {
        String readerUrl = onlineContent.getReaderUrl();

        if (!URLUtil.isValidUrl(readerUrl))
            throw new IllegalArgumentException("Invalid gallery URL : " + readerUrl);

        Timber.d("Gallery URL: %s", readerUrl);

        processedUrl = onlineContent.getGalleryUrl();
        EventBus.getDefault().register(this);

        List<ImageFile> result;
        try {
            List<String> imgUrls = parseImages(onlineContent);
            result = ParseHelper.urlsToImageFiles(imgUrls, StatusContent.SAVED, onlineContent.getCoverImageUrl(), null);
            ParseHelper.setDownloadParams(result, onlineContent.getSite().getUrl());
        } finally {
            EventBus.getDefault().unregister(this);
        }

        return result;
    }

    public List<ImageFile> parseChapterImageListImpl(@NonNull String url, @NonNull Content content) throws Exception {
        if (!URLUtil.isValidUrl(url))
            throw new IllegalArgumentException("Invalid gallery URL : " + url);

        Timber.d("Chapter URL: %s", url);

        processedUrl = url;
        EventBus.getDefault().register(this);

        List<ImageFile> result;
        try {
            List<String> imgUrls = parseImages(url, content.getDownloadParams(), null);
            result = ParseHelper.urlsToImageFiles(imgUrls, StatusContent.SAVED, null, null);
            ParseHelper.setDownloadParams(result, content.getSite().getUrl());
        } finally {
            EventBus.getDefault().unregister(this);
        }

        return result;
    }

    public Optional<ImageFile> parseBackupUrl(@NonNull String url, @NonNull Map<String, String> requestHeaders, int order, int maxPages, Chapter chapter) {
        // Default behaviour; this class does not use backup URLs
        ImageFile img = ImageFile.fromImageUrl(order, url, StatusContent.SAVED, maxPages);
        if (chapter != null) img.setChapter(chapter);
        return Optional.of(img);
    }

    public ImmutablePair<String, Optional<String>> parseImagePage(@NonNull String url, @NonNull List<Pair<String, String>> requestHeaders) throws IOException, LimitReachedException, EmptyResultException {
        throw new NotImplementedException("Parser does not implement parseImagePage");
    }

    @Override
    public String getAltUrl(@NonNull String url) {
        return "";
    }

    void progressStart(@NonNull Content onlineContent, @Nullable Content storedContent, int maxSteps) {
        if (progress.hasStarted()) return;
        long storedId = (storedContent != null) ? storedContent.getId() : -1;
        progress.start(onlineContent.getId(), storedId, maxSteps);
    }

    // TODO method to free resources after caller decides work's done

    void progressPlus() {
        progress.advance();
    }

    void progressComplete() {
        progress.complete();
    }

    /**
     * Download event handler called by the event bus
     *
     * @param event Download event
     */
    @Subscribe
    public void onDownloadCommand(DownloadCommandEvent event) {
        switch (event.getType()) {
            case EV_PAUSE:
            case EV_CANCEL:
            case EV_SKIP:
                processHalted.set(true);
                break;
            case EV_INTERRUPT_CONTENT:
                if (event.getContent() != null && event.getContent().getGalleryUrl().equals(processedUrl)) {
                    processHalted.set(true);
                    processedUrl = "";
                }
                break;
            case EV_UNPAUSE:
            default:
                // Other events aren't handled here
        }
    }

    protected List<Pair<String, String>> fetchHeaders(@NonNull Content content) {
        return fetchHeaders(content.getGalleryUrl(), content.getDownloadParams());
    }

    protected List<Pair<String, String>> fetchHeaders(@NonNull String url, String downloadParams) {
        List<Pair<String, String>> headers = new ArrayList<>();
        if (downloadParams != null) ParseHelper.addSavedCookiesToHeader(downloadParams, headers);
        headers.add(new Pair<>(HttpHelper.HEADER_REFERER_KEY, url));
        return headers;
    }
}

File: app/src/main/java/me/devsaki/hentoid/parsers/images/BaseImageListParser.kt
package me.devsaki.hentoid.parsers.images

import android.webkit.URLUtil
import androidx.core.util.Pair
import me.devsaki.hentoid.database.domains.Chapter
import me.devsaki.hentoid.database.domains.Content
import me.devsaki.hentoid.database.domains.ImageFile
import me.devsaki.hentoid.enums.StatusContent
import me.devsaki.hentoid.events.DownloadCommandEvent
import me.devsaki.hentoid.parsers.ParseHelper
import me.devsaki.hentoid.util.network.HttpHelper
import org.apache.commons.lang3.NotImplementedException
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import timber.log.Timber
import java.util.concurrent.atomic.AtomicBoolean

abstract class BaseImageListParser : ImageListParser {
    private val progress = ParseProgress()
    protected var processHalted = AtomicBoolean(false)
    protected var processedUrl = ""

    protected abstract fun isChapterUrl(url: String): Boolean

    @Throws(Exception::class)
    protected abstract fun parseImages(content: Content): List<String>

    @Throws(Exception::class)
    protected abstract fun parseImages(
        chapterUrl: String,
        downloadParams: String?,
        headers: List<Pair<String, String>>?
    ): List<String>

    override fun parseImageList(content: Content, url: String): List<ImageFile> {
        return if (isChapterUrl(url)) parseChapterImageListImpl(url, content)
        else parseImageListImpl(content, null)
    }

    override fun parseImageList(onlineContent: Content, storedContent: Content): List<ImageFile> {
        return parseImageListImpl(onlineContent, storedContent)
    }

    override fun parseImagePage(
        url: String,
        requestHeaders: List<Pair<String, String>>
    ): Pair<String, String?> {
        throw NotImplementedException("Parser does not implement parseImagePage")
    }

    override fun parseBackupUrl(
        url: String,
        requestHeaders: Map<String, String>,
        order: Int,
        maxPages: Int,
        chapter: Chapter?
    ): ImageFile? {
        // Default behaviour; this class does not use backup URLs
        val img = ImageFile.fromImageUrl(order, url, StatusContent.SAVED, maxPages)
        if (chapter != null) img.setChapter(chapter)
        return img
    }

    override fun getAltUrl(url: String): String {
        return ""
    }

    override fun clear() {
        // No default implementation
    }

    /**
     * Default implementation for sources that don't enrich content over time by publishing extra chapters or pages
     */
    @Throws(Exception::class)
    protected open fun parseImageListImpl(
        onlineContent: Content,
        storedContent: Content?
    ): List<ImageFile> {
        val readerUrl = onlineContent.readerUrl
        require(URLUtil.isValidUrl(readerUrl)) { "Invalid gallery URL : $readerUrl" }
        Timber.d("Gallery URL: %s", readerUrl)
        processedUrl = onlineContent.galleryUrl
        EventBus.getDefault().register(this)
        val result: List<ImageFile>
        try {
            val imgUrls = parseImages(onlineContent)
            result = ParseHelper.urlsToImageFiles(
                imgUrls,
                StatusContent.SAVED,
                onlineContent.coverImageUrl,
                null
            )
            ParseHelper.setDownloadParams(result, onlineContent.site.url)
        } finally {
            EventBus.getDefault().unregister(this)
        }
        return result
    }

    @Throws(Exception::class)
    protected open fun parseChapterImageListImpl(url: String, content: Content): List<ImageFile> {
        require(URLUtil.isValidUrl(url)) { "Invalid gallery URL : $url" }
        Timber.d("Chapter URL: %s", url)
        processedUrl = url
        EventBus.getDefault().register(this)
        val result: List<ImageFile>
        try {
            val imgUrls = parseImages(url, content.downloadParams, null)
            result = ParseHelper.urlsToImageFiles(imgUrls, StatusContent.SAVED, null, null)
            ParseHelper.setDownloadParams(result, content.site.url)
        } finally {
            EventBus.getDefault().unregister(this)
        }
        return result
    }

    open fun progressStart(onlineContent: Content, storedContent: Content?, maxSteps: Int) {
        if (progress.hasStarted()) return
        val storedId = storedContent?.id ?: -1
        progress.start(onlineContent.id, storedId, maxSteps)
    }

    open fun progressPlus() {
        progress.advance()
    }

    open fun progressComplete() {
        progress.complete()
    }

    /**
     * Download event handler called by the event bus
     *
     * @param event Download event
     */
    @Subscribe
    open fun onDownloadCommand(event: DownloadCommandEvent) {
        when (event.type) {
            DownloadCommandEvent.Type.EV_PAUSE, DownloadCommandEvent.Type.EV_CANCEL, DownloadCommandEvent.Type.EV_SKIP
            -> processHalted.set(true)

            DownloadCommandEvent.Type.EV_INTERRUPT_CONTENT ->
                if (event.content != null && event.content.galleryUrl == processedUrl) {
                    processHalted.set(true)
                    processedUrl = ""
                }

            DownloadCommandEvent.Type.EV_UNPAUSE -> {}
        }
    }

    protected open fun fetchHeaders(content: Content): List<Pair<String, String>> {
        return fetchHeaders(content.galleryUrl, content.downloadParams)
    }

    protected open fun fetchHeaders(
        url: String,
        downloadParams: String?
    ): List<Pair<String, String>> {
        val headers: MutableList<Pair<String, String>> = ArrayList()
        if (downloadParams != null) ParseHelper.addSavedCookiesToHeader(downloadParams, headers)
        headers.add(Pair(HttpHelper.HEADER_REFERER_KEY, url))
        return headers
    }
}